File: app/src/main/java/de/achimonline/kickassembler/acbg/runconfiguration/KickAssemblerSettingsEditor.java
package de.achimonline.kickassembler.acbg.runconfiguration;

import com.intellij.execution.ui.CommonProgramParametersPanel;
import com.intellij.execution.ui.DefaultJreSelector;
import com.intellij.execution.ui.JrePathEditor;
import com.intellij.openapi.fileChooser.FileChooserDescriptor;
import com.intellij.openapi.options.ConfigurationException;
import com.intellij.openapi.options.SettingsEditor;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.LabeledComponent;
import com.intellij.openapi.ui.TextFieldWithBrowseButton;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;

import static de.achimonline.kickassembler.acbg.properties.KickAssemblerProperties.message;

public class KickAssemblerSettingsEditor extends SettingsEditor<KickAssemblerRunConfiguration> {
    private JPanel panel;

    private LabeledComponent<TextFieldWithBrowseButton> kickAssemblerFile;
    private JrePathEditor jrePathEditor;
    private CommonProgramParametersPanel commonProgramParametersPanel;

    public KickAssemblerSettingsEditor(Project project) {
        FileChooserDescriptor kickAssemblerFileChooserDescriptor = new FileChooserDescriptor(true, false, false, true, false, false);
        kickAssemblerFileChooserDescriptor.setTitle(message("runconfiguration.main.file.browse.title"));
        kickAssemblerFileChooserDescriptor.setDescription(message("runconfiguration.main.file.browse.description"));

        kickAssemblerFile.getComponent().addBrowseFolderListener(kickAssemblerFileChooserDescriptor.getTitle(), kickAssemblerFileChooserDescriptor.getDescription(), project, kickAssemblerFileChooserDescriptor);
        jrePathEditor.setDefaultJreSelector(DefaultJreSelector.projectSdk(project));
    }

    @Override
    protected void resetEditorFrom(@NotNull KickAssemblerRunConfiguration kickAssemblerRunConfiguration) {
        kickAssemblerFile.getComponent().setText(kickAssemblerRunConfiguration.getKickAssemblerFile());
        jrePathEditor.setPathOrName(kickAssemblerRunConfiguration.getKickAssemblerSdkPathOrName(), true);
        commonProgramParametersPanel.reset(kickAssemblerRunConfiguration.getKickAssemblerProgramParameters());
    }

    @Override
    protected void applyEditorTo(@NotNull KickAssemblerRunConfiguration kickAssemblerRunConfiguration) throws ConfigurationException {
        kickAssemblerRunConfiguration.setKickAssemblerFile(kickAssemblerFile.getComponent().getText());
        kickAssemblerRunConfiguration.setKickAssemblerSdkPathOrName(jrePathEditor.getJrePathOrName());
        commonProgramParametersPanel.applyTo(kickAssemblerRunConfiguration.getKickAssemblerProgramParameters());
    }

    @NotNull
    @Override
    protected JComponent createEditor() {
        return panel;
    }
}

File: src/main/kotlin/de/achimonline/kickassembler/acbg/runconfiguration/KickAssemblerSettingsEditor.kt
package de.achimonline.kickassembler.acbg.runconfiguration

import com.intellij.execution.ui.CommonProgramParametersPanel
import com.intellij.openapi.fileChooser.FileChooserDescriptor
import com.intellij.openapi.options.SettingsEditor
import com.intellij.openapi.project.Project
import com.intellij.openapi.projectRoots.SdkType
import com.intellij.openapi.roots.ui.configuration.JdkComboBox
import com.intellij.openapi.roots.ui.configuration.projectRoot.ProjectSdksModel
import com.intellij.openapi.ui.TextFieldWithBrowseButton
import com.intellij.ui.dsl.builder.Align
import com.intellij.ui.dsl.builder.AlignX
import com.intellij.ui.dsl.builder.panel
import de.achimonline.kickassembler.acbg.properties.KickAssemblerProperties.message
import de.achimonline.kickassembler.acbg.sdk.KickAssemblerSdk
import javax.swing.JComponent

@Suppress("DialogTitleCapitalization")
class KickAssemblerSettingsEditor(val project: Project) : SettingsEditor<KickAssemblerRunConfiguration<*>>() {
    private val kickAssemblerFile = TextFieldWithBrowseButton()
    private var kickAssemblerSdk: JdkComboBox
    private var kickAssemblerProgramParameters: CommonProgramParametersPanel

    private val projectSdksModel = ProjectSdksModel()

    init {
        val kickAssemblerFileChooserDescriptor = FileChooserDescriptor(
            /* chooseFiles = */ true,
            /* chooseFolders = */ false,
            /* chooseJars = */ false,
            /* chooseJarsAsFiles = */ true,
            /* chooseJarContents = */ false,
            /* chooseMultiple = */ false
        )
        kickAssemblerFileChooserDescriptor.title = message("runconfiguration.main.file.browse.title")
        kickAssemblerFileChooserDescriptor.description = message("runconfiguration.main.file.browse.description")

        kickAssemblerFile.addBrowseFolderListener(
            /* title = */ kickAssemblerFileChooserDescriptor.title,
            /* description = */ kickAssemblerFileChooserDescriptor.description,
            /* project = */ project,
            /* fileChooserDescriptor = */ kickAssemblerFileChooserDescriptor
        )

        projectSdksModel.syncSdks()

        val kickAssemblerSdkType = SdkType.findInstance(KickAssemblerSdk::class.java)

        kickAssemblerSdk = JdkComboBox(
            /* project = */ project,
            /* sdkModel = */ projectSdksModel,
            /* sdkTypeFilter = */ { it is KickAssemblerSdk },
            /* sdkFilter = */ { it.sdkType == kickAssemblerSdkType },
            /* suggestedSdkFilter = */ { it.sdkType == kickAssemblerSdkType },
            /* creationFilter = */ null,
            /* onNewSdkAdded = */ null
        )

        kickAssemblerProgramParameters = CommonProgramParametersPanel()
    }

    override fun resetEditorFrom(kickAssemblerRunConfiguration: KickAssemblerRunConfiguration<*>) {
        kickAssemblerFile.text = kickAssemblerRunConfiguration.kickAssemblerFile
        kickAssemblerSdk.selectedJdk = KickAssemblerSdk.getSdkFromHomePathOrName(kickAssemblerRunConfiguration.kickAssemblerSdkPathOrName)
        kickAssemblerProgramParameters.reset(kickAssemblerRunConfiguration.kickAssemblerProgramParameters)
    }

    override fun applyEditorTo(kickAssemblerRunConfiguration: KickAssemblerRunConfiguration<*>) {
        kickAssemblerRunConfiguration.kickAssemblerFile = kickAssemblerFile.text
        kickAssemblerRunConfiguration.kickAssemblerSdkPathOrName = kickAssemblerSdk.selectedJdk?.homePath
        kickAssemblerProgramParameters.applyTo(kickAssemblerRunConfiguration.kickAssemblerProgramParameters)
    }

    override fun createEditor(): JComponent {
        return panel {
            group {
                row(message("runconfiguration.main.file")) {
                    cell(kickAssemblerFile)
                        .align(AlignX.FILL)
                }

                row(message("runconfiguration.sdk")) {
                    cell(kickAssemblerSdk)
                        .align(AlignX.FILL)
                }
            }

            group {
                row {
                    cell(kickAssemblerProgramParameters)
                        .resizableColumn()
                        .align(Align.FILL)
                }
            }
        }
    }
}
